File: src/012/p012.ts
const divide = (n: number, d: number) => {
  let exponent = 0;
  while (n % d === 0) {
    n /= d;
    exponent++;
  }
  return { remainder: n, exponent };
};

const factorize = (n: number) => {
  const factors: Record<number, number> = {};

  const factorizeInternal = (d: number) => {
    const { remainder, exponent } = divide(n, d);
    if (exponent > 0) {
      factors[d] = exponent;
      n = remainder;
    }
  };

  factorizeInternal(2);

  // Ideally, we should generate prime numbers, but just testing all odd numbers is fast enough
  const sqrt = Math.sqrt(n);
  for (let d = 3; n > 1 && d <= sqrt; d += 2) {
    factorizeInternal(d);
  }

  if (n > 1) {
    factors[n] = 1;
  }
  return factors;
};

const divisorCount = (n: number) => {
  return Object.values(factorize(n))
    .map((power) => power + 1)
    .reduce((prev, cur) => prev * cur, 1);
};

const triangleNumbers = function* () {
  for (let t = 0, n = 1; ; n++) {
    t += n;
    yield t;
  }
};

export const solve = (): string => {
  for (const t of triangleNumbers()) {
    if (divisorCount(t) > 500) {
      return t.toString();
    }
  }
  throw new Error("No solution found");
};

File: src/012/p012.py
import itertools
from src.utils.py.math import count_of_divisors
from src.utils.py.prime import generate_primes, factorize
from src.utils.py.series import triangle_numbers


def solve1() -> int:
    primes = generate_primes(9000)
    for triangle in triangle_numbers():
        factors = factorize(triangle, primes=primes)
        if count_of_divisors(triangle, factors=factors) > 500:
            return triangle
    raise Exception("No solution found")  # pylint: disable=broad-exception-raised


def solve() -> int:
    primes = generate_primes(120)
    for k in itertools.count(1):
        count1 = count_of_divisors(k, factors=factorize(k, primes=primes))

        factors = factorize(2 * k - 1, primes=primes)
        count2 = count_of_divisors(2 * k - 1, factors=factors)
        if count1 * count2 > 500:
            return k * (2 * k - 1)

        factors = factorize(2 * k + 1, primes=primes)
        count2 = count_of_divisors(2 * k + 1, factors=factors)
        if count1 * count2 > 500:
            return k * (2 * k + 1)
    raise Exception("No solution found")  # pylint: disable=broad-exception-raised
